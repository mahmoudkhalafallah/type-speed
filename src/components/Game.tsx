import React, { useState, useEffect } from 'react'
import styled from 'styled-components'
import InfoList from '../components/InfoList'
import Racer from '../components/Racer'
import { HISTORY_URL } from '../constants'

const GameOver = styled.h2`
transition: 0.2s ease;
color: #c90000;
font-size: 48px;
z-index: 5;
display: flex;
align-items: center;
justify-content: center;
`

const Game: React.FC = () => {
    const [gameOver, setGameOver] = useState(false)
    const [wpm, setWpm] = useState(0)
    const [completionPercent, setCompletionPercent] = useState(0)
    const [gameRecordsHistoryId, setGameRecordsHistoryId] = useState('')


    useEffect(() => {
        if (gameOver && completionPercent > 0) {
            fetch(HISTORY_URL,
                {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify({
                        wpm,
                        cp: completionPercent
                    })
                }
            ).then(res => res.json()).then((data: { uri: string }) => {
                const splittedText = data.uri.split('/')
                setGameRecordsHistoryId(splittedText[splittedText.length - 1])
            })
        }
    }, [gameOver, completionPercent, wpm])

    return (
        <div className="Game">
            {gameOver && <GameOver>Game Over</GameOver>}
            <InfoList gameOver={gameOver} endGame={setGameOver} wpm={wpm} cp={completionPercent} historyId={gameRecordsHistoryId} />
            <Racer gameOver={gameOver} endGame={setGameOver} setWpm={setWpm} setCp={setCompletionPercent} />
        </div>
    )
}

export default Game
