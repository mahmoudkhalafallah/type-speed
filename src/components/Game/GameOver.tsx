import React from 'react'
import styled from 'styled-components'

const GameOverTitle = styled.h2`
transition: 0.2s ease;
color: #c90000;
font-size: 40px;
z-index: 5;
text-align: center;
font-weight: 100;
font-family: "capture_it";
`

const PlayAgainBtn = styled.button`
display: block;
margin: 0 auto;
background: #000;
border: 0;
padding: 15px 20px;
color: #fcc944;
font-size: 20px;
cursor: pointer;
transition: 0.3s ease;
font-family: "capture_it";
&:hover {
    background: #1c9d1c;
    color: #fff;
    transform: scale(1.1);
}
`
const GameOver: React.FC<{ restartGame: (value: React.SetStateAction<boolean>) => void }> = ({ restartGame }) =>
    <>
        <GameOverTitle>Game Over</GameOverTitle>
        <PlayAgainBtn onClick={() => { restartGame(true) }}>Play Again !</PlayAgainBtn>
    </>
export default GameOver


