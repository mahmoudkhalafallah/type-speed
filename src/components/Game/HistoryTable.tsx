import React from 'react'
import styled from 'styled-components'
// eslint-disable-next-line no-unused-vars
import { WindowLocation } from '@reach/router'

const TableContainer = styled.section`
background: #fff;
padding: 20px;
margin: 50px 0;
.section-header {
    margin: 0;
}
`

const Table = styled.table`
    width: 100%;
    margin-top: 25px;
    th, td {
        text-align: start;
        padding: 10px 5px;
    }
    tbody tr:nth-child(odd) {
        background: #eee;
    }
`

const HistoryTable: React.FC<{ history: any[] }> = ({ history }) =>
    <TableContainer>
        <h2 className='section-header'>History</h2>
        <Table>
            <thead>
                <tr>
                    <th />
                    <th>Date</th>
                    <th>WPM</th>
                    <th>Completion Percent</th>
                </tr>
            </thead>
            <tbody>
                {history.map(
                    (h: any, index: number) => {
                        const date = new Date(+h.date)
                        return <tr key={h.date}>
                            <td>{index + 1}</td>
                            <td>{`
                        ${date.getMonth()}-${date.getDate()}-${date.getFullYear()}
                         ${date.getHours()}:${date.getMinutes()}
                        `}</td>
                            <td>{h.wpm}</td>
                            <td>{h.cp}</td>
                        </tr>
                    }
                )}
            </tbody>
        </Table>
    </TableContainer>
export default HistoryTable