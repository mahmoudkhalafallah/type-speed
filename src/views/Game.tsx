import React, { useState, useEffect } from 'react'
import InfoList from '../components/InfoList'
import Racer from '../components/Racer'
import { HISTORY_URL } from '../constants'
import { RouteComponentProps } from "@reach/router" // eslint-disable-line no-unused-vars
import StatsContext from '../utils/context/StatsContext'
import GameOver from '../components/Game/GameOver'
import HistoryTable from '../components/Game/HistoryTable'


const Game: React.FC<RouteComponentProps> = ({ location }) => {
    const [gameOver, setGameOver] = useState(false)
    const [restartGame, setRestartGame] = useState(false)
    const [wpm, setWpm] = useState(0)
    const [cp, setCp] = useState(0)
    const [historyId, setHistoryId] = useState('')

    useEffect(() => {
        if (gameOver && cp > 0) {
            fetch(HISTORY_URL,
                {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify({
                        wpm,
                        cp: cp
                    })
                }
            ).then(res => res.json()).then((data: { uri: string }) => {
                const splittedText = data.uri.split('/')
                setHistoryId(splittedText[splittedText.length - 1])
            })
                .catch((err) => {
                    console.log(err)
                })
        }
    }, [gameOver, cp, wpm])


    useEffect(() => {
        if (restartGame) {
            setGameOver(false)
            setRestartGame(false)
        }
    }, [restartGame])

    return (
        <>
            {gameOver && <GameOver restartGame={setRestartGame} />}
            {
                !restartGame &&
                <StatsContext.Provider value={{ wpm, cp, historyId }}>
                    <InfoList gameOver={gameOver} endGame={setGameOver} />
                    <Racer gameOver={gameOver} endGame={setGameOver} setWpm={setWpm} setCp={setCp} />
                </StatsContext.Provider>
            }
            {
                (location && location.state && location.state.history) ? <HistoryTable history={location.state.history} /> : null
            }
        </>
    )
}

export default Game
