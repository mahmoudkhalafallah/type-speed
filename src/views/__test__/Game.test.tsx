import React from 'react'
import Game from '../Game'
import renderer from 'react-test-renderer'
import 'jest-fetch-mock'

test('Game rendered correctly', () => {
    const component = renderer.create(
        <Game />,
    )
    let tree = component.toJSON()

    expect(tree).toMatchSnapshot()
})
