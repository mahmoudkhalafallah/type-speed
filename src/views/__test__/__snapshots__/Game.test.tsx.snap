// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game rendered correctly 1`] = `
<div
  className="Game"
>
  <ul
    className="sc-bdVaJa cSScn"
  >
    <li
      className="sc-bwzfXH fwsDhg"
    >
      Time:
      <span>
        3
        :
        00
      </span>
    </li>
    <li
      className="sc-bwzfXH fwsDhg"
    >
      Words Per Minute:
      <span
        className="sc-htpNat grLPkg"
      >
        0
         WPM
      </span>
    </li>
  </ul>
  <section
    className="sc-dnqmqq hoMqEr"
  />
</div>
`;
